File: plugin/plugins/flurry/proj.android/src/org/cocos2dx/plugin/AnalyticsFlurry.java
/****************************************************************************
Copyright (c) 2012-2013 cocos2d-x.org

http://www.cocos2d-x.org

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
package org.cocos2dx.plugin;

import java.util.Hashtable;
import java.util.Iterator;

import org.json.JSONObject;

import android.content.Context;
import android.util.Log;

import com.flurry.android.Constants;
import com.flurry.android.FlurryAgent;

public class AnalyticsFlurry implements InterfaceAnalytics {

    private Context mContext = null;
    protected static String TAG = "AnalyticsFlurry";

    protected static void LogE(String msg, Exception e) {
        Log.e(TAG, msg, e);
        e.printStackTrace();
    }

    private static boolean isDebug = false;
    protected static void LogD(String msg) {
        if (isDebug) {
            Log.d(TAG, msg);
        }
    }

    public AnalyticsFlurry(Context context) {
        mContext = context;
    }
    
    @Override
    public void startSession(String appKey) {
        LogD("startSession invoked!");
        final String curKey = appKey;
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                try {
                    Class.forName("android.os.AsyncTask");
                } catch (ClassNotFoundException e) {
                   e.printStackTrace();
                }
                FlurryAgent.onStartSession(mContext, curKey);
            }
            
        });
    }

    @Override
    public void stopSession() {
        LogD("stopSession invoked!");
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                FlurryAgent.onEndSession(mContext);
            }
        });
    }

    @Override
    public void setSessionContinueMillis(int millis) {
        LogD("setSessionContinueMillis invoked!");
        final int curMillis = millis;
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                FlurryAgent.setContinueSessionMillis(curMillis);
            }
        });
    }

    @Override
    public void setCaptureUncaughtException(boolean isEnabled) {
        LogD("setCaptureUncaughtException invoked!");
        final boolean curEnable = isEnabled;
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                FlurryAgent.setCaptureUncaughtExceptions(curEnable);
            }
        });
    }

    @Override
    public void setDebugMode(boolean isDebugMode) {
        isDebug = isDebugMode;
        final boolean curDebugMode = isDebug;
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                FlurryAgent.setLogEnabled(curDebugMode);
                if (curDebugMode) {
                    FlurryAgent.setLogLevel(Log.DEBUG);
                }
            }
        });
    }

    @Override
    public void logError(String errorId, String message) {
        LogD("logError invoked!");
        final String curID = errorId;
        final String curMsg = message;
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                FlurryAgent.onError(curID, curMsg, "");
            }
        });
    }

    @Override
    public void logEvent(String eventId) {
        LogD("logEvent(eventId) invoked!");
        final String curId = eventId;
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                FlurryAgent.logEvent(curId);
            }
        });
    }

    @Override
    public void logEvent(String eventId, Hashtable<String, String> paramMap) {
        LogD("logEvent(eventId, paramMap) invoked!");
        final String curId = eventId;
        final Hashtable<String, String> curParam = paramMap;
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                FlurryAgent.logEvent(curId, curParam);
            }
        });
    }

    @Override
    public void logTimedEventBegin(String eventId) {
        LogD("logTimedEventBegin invoked!");
        final String curId = eventId;
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                FlurryAgent.logEvent(curId, true);
            }
        });
    }

    @Override
    public void logTimedEventEnd(String eventId) {
        LogD("logTimedEventEnd invoked!");
        final String curId = eventId;
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                FlurryAgent.endTimedEvent(curId);
            }
        });
    }

    @Override
    public String getSDKVersion() {
        return "3.2.1";
    }

    protected void logTimedEventBeginWithParams(JSONObject eventInfo) {
        LogD("logTimedEventBegin invoked!");
        final JSONObject curInfo = eventInfo;
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                try{
                    String eventId = curInfo.getString("Param1");
                    
                    if (curInfo.has("Param2"))
                    {
                        JSONObject params = curInfo.getJSONObject("Param2");
                        @SuppressWarnings("rawtypes")
                        Iterator it = params.keys();
                        Hashtable<String, String> paramMap = new Hashtable<String, String>();
                        while (it.hasNext()) {
                            String key = (String) it.next();
                            String value = params.getString(key);
                            paramMap.put(key, value);
                        }
                        FlurryAgent.logEvent(eventId, paramMap, true);
                    } else {
                        FlurryAgent.logEvent(eventId, true);
                    }
                } catch(Exception e){
                    LogE("Exception in logTimedEventBegin", e);
                }
            }
        });
    }
    
    protected void setReportLocation(boolean enabled) {
        LogD("setReportLocation invoked!");
        final boolean curEnable = enabled;
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                try{
                    FlurryAgent.setReportLocation(curEnable);
                } catch(Exception e){
                    LogE("Exception in setReportLocation", e);
                }
            }
        });
    }
    
    protected void  logPageView() {
        LogD("logPageView invoked!");
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                try{
                    FlurryAgent.onPageView();
                } catch(Exception e){
                    LogE("Exception in logPageView", e);
                }
            }
        });
    }

    protected void setVersionName(String versionName) {
        LogD("setVersionName invoked!");
        final String curVer = versionName;
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                try {
                    FlurryAgent.setVersionName(curVer);
                } catch(Exception e){
                    LogE("Exception in setVersionName", e);
                }
            }
        });
    }
    
    protected void setAge(int age) {
        LogD("setAge invoked!");
        final int curAge = age;
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                try {
                    FlurryAgent.setAge(curAge);
                } catch(Exception e){
                    LogE("Exception in setAge", e);
                }
            }
        });
    }
    
    protected void setGender(int gender) {
        LogD("setGender invoked!");
        final int curGender = gender;
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                try {
                    byte bGender;
                    if (1 == curGender) {
                        bGender = Constants.MALE;
                    } else {
                        bGender = Constants.FEMALE;
                    }
                    FlurryAgent.setGender(bGender);
                } catch(Exception e){
                    LogE("Exception in setGender", e);
                }
            }
        });
    }
    
    protected void setUserId(String userId) {
        LogD("setUserId invoked!");
        final String curUser = userId;
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                try {
                    FlurryAgent.setUserId(curUser);
                } catch(Exception e){
                    LogE("Exception in setUserId", e);
                }
            }
        });
    }
    
    protected void setUseHttps(boolean useHttps) {
        LogD("setUseHttps invoked!");
        
        final boolean curCfg = useHttps;
        PluginWrapper.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                try {
                    FlurryAgent.setUseHttps(curCfg);
                } catch(Exception e){
                    LogE("Exception in setUseHttps", e);
                }
            }
        });
    }

    @Override
    public String getPluginVersion() {
        return "0.2.0";
    }
}

File: plugin/plugins/flurry/proj.ios/AnalyticsFlurry.m
/****************************************************************************
Copyright (c) 2012-2013 cocos2d-x.org

http://www.cocos2d-x.org

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
#import "AnalyticsFlurry.h"
#import "Flurry.h"

#define OUTPUT_LOG(...)     if (self.debug) NSLog(__VA_ARGS__);

@implementation AnalyticsFlurry

@synthesize debug = __debug;

- (void) startSession: (NSString*) appKey
{
    [Flurry startSession:appKey];
}

- (void) stopSession
{
    OUTPUT_LOG(@"Flurry stopSession in flurry not available on iOS");
}

- (void) setSessionContinueMillis: (long) millis
{
    OUTPUT_LOG(@"Flurry setSessionContinueMillis invoked(%ld)", millis);
    int seconds = (int)(millis / 1000);
    [Flurry setSessionContinueSeconds:seconds];
}

- (void) setCaptureUncaughtException: (BOOL) isEnabled
{
    OUTPUT_LOG(@"Flurry setCaptureUncaughtException in flurry not available on iOS");
}

- (void) setDebugMode: (BOOL) isDebugMode
{
    OUTPUT_LOG(@"Flurry setDebugMode invoked(%d)", isDebugMode);
    self.debug = isDebugMode;
    [Flurry setDebugLogEnabled:isDebugMode];
}

- (void) logError: (NSString*) errorId withMsg:(NSString*) message
{
    OUTPUT_LOG(@"Flurry logError invoked(%@, %@)", errorId, message);
    NSString* msg = nil;
    if (nil == message) {
        msg = @"";
    } else {
        msg = message;
    }
    [Flurry logError:errorId message:msg exception:nil];
}

- (void) logEvent: (NSString*) eventId
{
    OUTPUT_LOG(@"Flurry logEvent invoked(%@)", eventId);
    [Flurry logEvent:eventId];
}

- (void) logEvent: (NSString*) eventId withParam:(NSMutableDictionary*) paramMap
{
    OUTPUT_LOG(@"Flurry logEventWithParams invoked (%@, %@)", eventId, [paramMap debugDescription]);
    [Flurry logEvent:eventId withParameters:paramMap];
}

- (void) logTimedEventBegin: (NSString*) eventId
{
    OUTPUT_LOG(@"Flurry logTimedEventBegin invoked (%@)", eventId);
    [Flurry logEvent:eventId timed:YES];
}

- (void) logTimedEventEnd: (NSString*) eventId
{
    OUTPUT_LOG(@"Flurry logTimedEventEnd invoked (%@)", eventId);
    [Flurry endTimedEvent:eventId withParameters:nil];
}

- (NSString*) getSDKVersion
{
    return @"4.2.1";
}

- (NSString*) getPluginVersion
{
    return @"0.2.0";
}

- (void) setAge: (NSNumber*) age
{
    int nAge = [age integerValue];
    OUTPUT_LOG(@"Flurry setAge invoked (%d)", nAge);
    [Flurry setAge:nAge];
}

- (void) setGender: (NSNumber*) gender
{
    OUTPUT_LOG(@"Flurry setGender invoked (%@)", gender);
    int nValue = [gender intValue];
    NSString* strGender;
    if (nValue == 1) {
        strGender = @"m";
    } else {
        strGender = @"f";
    }
    [Flurry setGender:strGender];
}

- (void) setUserId: (NSString*) userId
{
    OUTPUT_LOG(@"Flurry setUserId invoked (%@)", userId);
    [Flurry setUserID:userId];
}

- (void) setUseHttps: (NSNumber*) enabled
{
    BOOL bEnabled = [enabled boolValue];
    OUTPUT_LOG(@"Flurry setUseHttps invoked (%@)", enabled);
    [Flurry setSecureTransportEnabled:bEnabled];
}

- (void) logPageView
{
    OUTPUT_LOG(@"Flurry logPageView invoked");
    [Flurry logPageView];
}

- (void) setVersionName: (NSString*) versionName
{
    OUTPUT_LOG(@"Flurry setVersionName invoked (%@)", versionName);
    [Flurry setAppVersion:versionName];
}

- (void) logTimedEventBeginWithParams: (NSMutableDictionary*) paramMap
{
    OUTPUT_LOG(@"Flurry logTimedEventBeginWithParams invoked (%@)", [paramMap debugDescription]);
    NSString* eventId = (NSString*) [paramMap objectForKey:@"Param1"];
    NSMutableDictionary* params = (NSMutableDictionary*) [paramMap objectForKey:@"Param2"];
    if (params) {
        [Flurry logEvent:eventId withParameters:paramMap timed:YES];
    } else {
        [Flurry logEvent:eventId timed:YES];
    }
}

- (void) logTimedEventEndWithParams: (NSMutableDictionary*) paramMap
{
    OUTPUT_LOG(@"Flurry logTimedEventEndWithParams invoked (%@)", [paramMap debugDescription]);
    NSString* eventId = (NSString*) [paramMap objectForKey:@"Param1"];
    NSMutableDictionary* params = (NSMutableDictionary*) [paramMap objectForKey:@"Param2"];
    [Flurry endTimedEvent:eventId withParameters:params];
}

@end
